nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+    return hubClient;
+  }
+
+  private pairWithService() {
+    console.info('Starting pairing operation.');
+    const pairingRequestPatch = {
+      thief: {
+        pairing: {
+          requestedServicePool: requestedServicePool,
+          serviceInstanceId: null,
+          runId: runId
+        }
+      }
+    };
+
+    const pairWithServiceAttemptRecursive = () => {
+      return new Promise<void>((res, rej) => {
+        console.info('Updating pairing reported props: %j', pairingRequestPatch);
+        promisify(this.twin.properties.reported.update)(pairingRequestPatch)
+          .then(() => {
+            // Register an event listener to listen for response from service app
+            this.twin.on('properties.desired.thief.pairing', received => {
+              if (this.serviceInstanceId) {
+                console.info('Already paired. Ignoring');
+              }
+              else if (!received.runId || !received.serviceInstanceId) {
+               console.info('runId and/or serviceId is missing. Ignoring.');
+              }
+              else if (received.runId !== runId) {
+                console.info(`runId mismatch. Ignoring. (received ${received.runId}, expected ${runId})`);
+              }
+              else {
+                console.info(`Service app ${received.serviceInstanceId} claimed this device instance`);
+                this.twin.removeAllListeners('properties.desired.thief.pairing');
+                clearTimeout(this.pairingAttemptTimer);
+                this.serviceInstanceId = received.serviceInstanceId;
+                const pairingAcceptPatch = {
+                  thief: {
+                    pairing: {
+                      serviceInstanceId: received.serviceInstanceId,
+                      runId: runId
+                    }
+                  }
+                };
+                console.info('Updating pairing reported props: %j', pairingAcceptPatch);
+                promisify(this.twin.properties.reported.update)(pairingAcceptPatch)
+                  .then(res)
+                  .catch(e => rej(new Error('Updating reported properties failed: ' + e.message)));
+              }
+            });
+
+            // If it's taking too long for a service app to respond to our pairing request,
+            // retry by recursively calling pairWithServiceAttemptRecursive()
+            this.pairingAttemptTimer = setTimeout(() => {
+              console.info('Pairing response timeout. Requesting again.');
+              this.twin.removeAllListeners('properties.desired.thief.pairing');
+              pairWithServiceAttemptRecursive().then(res).catch(e => rej(e));
+            }, 1000 * this.settings.pairingRequestSendIntervalInSeconds);
+          })
+          .catch(e => rej(new Error('Updating reported properties failed: ' + e.message)));
+      });
+    }
+
+    return new Promise<void>((res, rej) => {      
```

Comment:
>(res, rej) [](start = 29, length = 10)

We should follow the same convention as the Node team, but I would prefer `resolve, reject` instead of `res, rej`.